lib/pbench/test/unit/server/test_sync.py

```diff
@@ -1,150 +1,239 @@
+
     def test_construct(self, make_logger):
         """A few simple checks on the sync constructor, including the
         string conversion.
         """
-        sync = Sync(make_logger, "test")
+        sync = Sync(make_logger, OperationName.UPLOAD)
         assert sync.logger is make_logger
-        assert sync.component == "test"
-        assert str(sync) == "<Synchronizer for component 'test'>"
+        assert sync.component == OperationName.UPLOAD
+        assert str(sync) == "<Synchronizer for component 'UPLOAD'>"
 
-    def test_next(self, make_logger, more_datasets):
-        """Test that the sync next operation returns the datasets enabled for
-        the requested operation
-        """
+    def test_error(self, make_logger, more_datasets):
+        """Test that the sync error operation writes the expected data."""
         drb = Dataset.query(name="drb")
-        fio_1 = Dataset.query(name="fio_1")
-        fio_2 = Dataset.query(name="fio_2")
-        sync = Sync(make_logger, "test")
-        Metadata.setvalue(drb, Metadata.OPERATION, ["UNPACK"])
-        Metadata.setvalue(fio_1, Metadata.OPERATION, ["UNPACK"])
-        Metadata.setvalue(fio_2, Metadata.OPERATION, ["BACKUP"])
-        list = sync.next(Operation.UNPACK)
-        assert ["drb", "fio_1"] == sorted(d.name for d in list)
-
-    def test_next_failure(self, monkeypatch, make_logger):
-        """Test the behavior of the sync next behavior when a DB failure
-        occurs.
-        """
-
-        def fake_query(self, *entities, **kwargs):
-            raise Exception("nothing happened")
-
-        sync = Sync(make_logger, "test")
-        monkeypatch.setattr("pbench.server.sync.Database.db_session.query", fake_query)
-        with pytest.raises(SyncSqlError):
-            sync.next(Operation.UNPACK)
+        sync = Sync(make_logger, OperationName.UPLOAD)
+        sync.update(drb, enabled=[OperationName.UPLOAD])
+        assert Metadata.getvalue(drb, "dataset.operations") == {
+            "UPLOAD": {"state": "READY", "message": None}
+        }
+        sync.error(drb, "this is an error")
+        assert Metadata.getvalue(drb, "dataset.operations") == {
+            "UPLOAD": {"state": "FAILED", "message": "this is an error"}
+        }
 
-    def test_error(self, make_logger, more_datasets):
-        """Test that the sync error operation writes the expected metadata."""
+    def test_error_new(self, make_logger, more_datasets):
+        """Test that the sync error operation creates an operation row if one
+        didn't already exist."""
         drb = Dataset.query(name="drb")
-        sync = Sync(make_logger, "test")
+        sync = Sync(make_logger, OperationName.UPLOAD)
         sync.error(drb, "this is an error")
-        assert Metadata.getvalue(drb, "server.status.test") == "this is an error"
+        assert Metadata.getvalue(drb, "dataset.operations") == {
+            "UPLOAD": {"state": "FAILED", "message": "this is an error"}
+        }
```

Comment:
Did the operation row already exist when the `Sync.error()` call was made?  (There's no assertion to check....)

Reply:
I suppose I could add a validation... it didn't seem worthwhile at the time since there are a lot of others.